### PR TITLE
Add ignore list capability (Implements #21)

### PR DESCRIPTION
**Pull request by dvdchr**
Still a work in progress, but this implements #21. 
(This is my first contribution for macOS apps, so please bear with me! :-)

Here's the gist of the idea:
- Maintain a list of apps to be excluded/ignored on every `tick` calls (currently I call it a "whitelist", but maybe there's a more suitable name?)
- We can get a list of running applications (via `NSRunningApplication`) and loop over each ones to check their `active` property. Once we get the active application name, we can do a search against our app list. If it exists in our list then do nothing (this happens as early as possible within the `tick` method).
- The list can be persisted in user defaults and loaded everytime the app starts.

Here's a preview of the whitelist sheet:
screen shot 2017-10-29 at 00 34 41

Let me know what you think! Any thoughts or comments are very much appreciated. 
Thanks!

### CODE REVIEW

**Comment by anishathalye** [comment]
This looks pretty good, thanks for the contribution!

What do you think is the right behavior for whitelisted vs non-whitelisted apps?

E.g. should `tick()` only do something if the app is whitelisted? In that case, what happens if we switch to a whitelisted app, the brightness changes, and we switch back to a non-whitelisted app?

**Comment by dvdchr** [comment]
@anishathalye Glad to help!

Sorry, maybe "whitelist" is not a correct term – I meant whitelist more like "exclusion list" or "ignore list" where apps listed here will not have its lightness monitored when it's on foreground. 

When we switch to an excluded app, the brightness will be preserved (i.e. brightness will not be increased when the excluded app is darker) and user would have to manually adjust the brightness.

What do you think?

**Comment by anishathalye** [comment]
That seems reasonable, and it should be easy to implement. In the tick handler, we can check if the foreground app is whitelisted, and if so, just not do anything.

**Comment by dvdchr** [comment]
Thanks for your input @anishathalye. Sorry I've been a bit busy these past weeks, will try to update the PR again by next weekend. 

Cheers!

**Comment by andrewmccullough** [comment]
Any progress on this? Just found the app and am loving using it!

**Comment by cheesemaker7651** [comment]
Thank you for working on this! I do visual effects for film and television and work in a low light controlled environment. Lumen has been saving my eyes, but being able to whitelist my main visual effects app would change my life! (as I'm currently forever turning lumen on and off manually)  I'm happy to donate money to this project!

**Comment by dvdchr** [comment]
Sorry I've been caught up with work these past few months.. Just wanted to update that I am (finally) committing some time to finish this feature! 😅

**Comment by anishathalye** [comment]
Great to hear -- looking forward to it :)

**Comment by dvdchr** [comment]
@anishathalye This is pretty much done, and would love to hear your thoughts on this (and probably give it a try :)

Note: I haven't tested this with multiple screen setup (I don't have one 😅)

File: Lumen/BrightnessController.m [hunk 1]
```diff
@@ -151,4 +188,17 @@ - (void)setBrightness:(float)level {
...
+#pragma mark - NSNotification Responder
+
+- (void)ignoreListChanged:(NSNotification *)notification {
+    self.updatingIgnoreList = YES;
+    
+    NSArray<NSString *> *newIgnoreList = notification.userInfo[@"updatedList"];
+    if (newIgnoreList) {
+        self.ignoredApplications = newIgnoreList;
+    }
+
```
Comment (anishathalye) [inline Lumen/BrightnessController.m]: There are a bunch of whitespace errors (you can see them by looking at `git diff --check master`)

File: Lumen/Ignore List/IgnoreListWindowController.xib [hunk 2]
```diff
@@ -0,0 +1,161 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (anishathalye) [inline Lumen/Ignore List/IgnoreListWindowController.xib]: Could we flatten this and move everything from the `Ignore List` folder into the `Lumen/` folder?

File: Lumen/Ignore List/IgnoreListWindowController.h [hunk 3]
```diff
@@ -0,0 +1,13 @@
+//
+//  IgnoreListWindowController.h
+//  Lumen
+//
+//  Created by David Christiandy on 28/10/17.
+//  Copyright © 2017 Anish Athalye. All rights reserved.
```
Comment (anishathalye) [inline Lumen/Ignore List/IgnoreListWindowController.h]: Could you follow the copyright notice on all the other files?

File: Lumen/Ignore List/IgnoreListWindowController.m [hunk 4]
```diff
@@ -0,0 +1,210 @@
+//
+//  IgnoreListWindowController.m
+//  Lumen
+//
+//  Created by David Christiandy on 28/10/17.
```
Comment (anishathalye) [inline Lumen/Ignore List/IgnoreListWindowController.m]: Could you follow the copyright notice on all the other files?

**Review by anishathalye — CHANGES_REQUESTED** [review]
Looks pretty good - left some minor comments on style and stuff.

Also, the ignore list has "+" and "-" buttons, and it seems that the last clicked button stays highlighted in blue; can we get rid of that behavior? It seems nicer if both buttons just stay the default grey.

A higher level point, consider the following scenario. I have three apps running, Terminal (dark), Mail (light), and Chrome. Suppose I put Chrome on the ignore list. Now, if I'm switching between apps, and I go Terminal -> Chrome, my backlight will be very bright, but if I go Mail -> Chrome, it'll be dark. It introduces this weird hysteresis. Given the original motivating use case mentioned in #21, maybe this is fine, but it's something that might be worth thinking about to see if we can come up with a clean solution to this problem. For example, for apps in the Ignore List, would it be better to keep track of brightness per-application (e.g. remembering the brightness the user wanted for Chrome) rather than just ignoring the app?

**Comment by dvdchr** [comment]
Thanks for the review! I'll try to fix them within this week. 

> For example, for apps in the Ignore List, would it be better to keep track of brightness per-application (e.g. remembering the brightness the user wanted for Chrome) rather than just ignoring the app?

I've given it some thought about this, and one immediate problem is if the user changes the brightness level manually, then Lumen will then try to adjust the brightness level back to when it was saved previously. If we update the persisted brightness level for the ignored app, that means we're back to square one: the brightness level will be the same as the previous unignored app.

I'm thinking, maybe we can differentiate whether the brightness adjustment is intended by the user, or it's actually from caused by Lumen adjusting the brightness in another app –– by checking on the **last focused/active application**.

- Using your example, when the user switched from Terminal to an ignored Chrome, we would know on the next `tick` that the last active app is Terminal, and Lumen would adjust the brightness back (to the persisted brightness level for Chrome). 
- Conversely, if the user adjusted the brightness level while an ignored Chrome is active – we would know that the last active app is still Chrome and it should be safe to assume that the change is user-intended.

I'm still thinking about other cases, though. What do you think?

**Comment by dvdchr** [comment]
Okay, so with the last commit, I've implemented the "remember" feature. I also made some changes – some important ones to note:

1. Inline static variables are converted to instance `@property` (i.e. `noticed`, `lastNoticed`).
2. Added `IgnoreListController` to wrap the logic behind managing the list and subscribing to changes. The motivation for adding this class is similar to `NSDocumentController`.
3. The logic of detecting whether the user manually changes the brightness is slightly modified, to accommodate for adjustments that are made by Lumen for "remembered brightness level". In summary, the logic modification is basically to **ignore brightness changes while an ignored app is in foreground** – whether the change is made by Lumen or by user.

I am still checking to see if I've missed any cases. Meanwhile, I'd appreciate your thoughts on whether the changes on existing behavior are good (especially no 3). 

Edit: I also forgot about the button state thing. Will adjust them later.

**Comment by anishathalye** [comment]
Yes, I think (3) is a good idea (and the other changes look good too).

I'm a bit busy for the next two weeks, but I should be able to take a closer look and give more detailed feedback afterwards. Looking forward to getting this feature merged soon.

**Comment by Sija** [comment]
@anishathalye Hi, this PR looks like pretty much ready, anything else to do here before merging?

**Comment by anishathalye** [comment]
@Sija thanks for the reminder.

Squashed in 1052dc1e47c7ae3d73b6717e33dd37bb04f2f38b, merged in 1dfe3aeceaef25e459d811d7ff4614ab8acecb01, and released in ebc95770feea072b749bc162a2d4fa279aa8879a. Thank you for the contribution, @davidchristiandy!

**Pull request closed without merging**